### ISSUE TITLE
Arbitrary late type support

### ISSUE THREAD

**Issue by airhorns** [xaviergonz/mobx-keystone#424]
For big projects that have a number of circular references, `types.model(() => SomeType)` is great for breaking the cycle and allowing types to reference themselves by not strictly requiring them to do so at require time. 

mobx-state-tree supports `types.late(() => ...)` for arbitrary late type support such that you can late-reference an enum type or a union type that maybe has model references within it that cause circular import issues.

Is this something you fine folks have considered for keystone, and would it be possible? Happy to look into it but would love to know if there are any hard blockers.

**Comment by xaviergonz** [comment]
recursive and cross referenced models are supported, see for example the test cases in https://github.com/xaviergonz/mobx-keystone/blob/master/packages/lib/test/model/recursive.test.ts

**Comment by xaviergonz** [comment]
the only requirement is that if you use types.model for those cases you have to include the type in the generic, like ```types. model<M>(() =>M) ```

**Comment by airhorns** [comment]
Right, that's working well for me, but as far as I can tell you can't do the same thing with a union of models that then cross reference / recurse, because the union doesn't accept a function, it only accepts already defined values. MST's API allows this because you can do 

```javascript
const outer = types.union(
  types.late(() => types.model({thing: outer}), 
  types.late(() => types.model({otherThing: outer}))
)
```

or what have you. My question is would you be open to adding something like `types.late` to mobx-keystone, or extending the other type definition things to accept functions instead of concrete types the same way `types.model` and `types.object` do?

**Comment by xaviergonz** [comment]
did you try inheritance? https://mobx-keystone.js.org/class-models#inheritance and/or the factory pattern https://mobx-keystone.js.org/class-models#factory-pattern--generics ?
